BABEL-3747 Bigint/Int/Smallint/Tinyint Degrees Function

pratikzode opened the pull request:
This Jira is a part of BABEL-736.
At first in Babel, the Transact-SQL functions degrees always return float instead of returning the same type in SQL server.
In this Jira we have added Bigint, Int, Smallint and Tinyint to degrees function.
Have added the test case to show Bigint, Int, Smallint and Tinyint to degrees.
Now this 3 is giving same output as SQL server.

Task: BABEL-3747
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 DEGREES, RADIANS and POWER return float always

In Babel, the Transact-SQL functions DEGREES, RADIANS and POWER always
return float instead of returning the same type.
Now we are returning power in int.

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
DEGREES, RADIANS and POWER return float always

In Babel, the Transact-SQL functions DEGREES, RADIANS and POWER always return float instead of returning the same type.
Now Transact-SQL functions DEGREES, RADIANS and POWER is returning the
same type as in case sql server.

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
DEGREES, RADIANS and POWER return float always
In Babel, the Transact-SQL functions DEGREES, RADIANS and POWER
 always return float instead of returning the same type in case if sql function.
Now DEGREES, RADIANS and POWER is returning the same type in case of babel.

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

xhfanhe commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1191,61 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+
+Datum
+int_power(PG_FUNCTION_ARGS)
+ {
```
Please fix the indentation

pratikzode replied:
Did the indentation changes

macwoody replied:
please double check all the indentations

xhfanhe commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -3174,3 +3183,7 @@ END;
+
+GRANT EXECUTE ON FUNCTION sys.power(int,int) TO PUBLIC;
+GRANT EXECUTE ON FUNCTION sys.radians(int) TO PUBLIC;
+GRANT EXECUTE ON FUNCTION sys.degrees(int) TO PUBLIC;
```
Please put these GRANT statements right after each CREATE.

pratikzode replied:
Did the modifications

xhfanhe commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -3814,11 +3814,17 @@ $$
 LANGUAGE 'pltsql';
 GRANT EXECUTE on PROCEDURE sys.sp_helpuser TO PUBLIC;
 
--- Identity related helper functions are no longer needed
-ALTER FUNCTION sys.get_min_id_from_table RENAME TO get_min_id_from_table_deprecated_in_2_3_0;
-ALTER FUNCTION sys.get_max_id_from_table RENAME TO get_max_id_from_table_deprecated_in_2_3_0;
-CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_min_id_from_table_deprecated_in_2_3_0');
-CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_max_id_from_table_deprecated_in_2_3_0');
```
These lines shouldn't be deleted

pratikzode replied:
Changed it again

xhfanhe commented on test/JDBC/expected/BABEL-736.out:
```diff
@@ -0,0 +1,69 @@
+GO
+~~START~~
+float
+5.72843209301375E9
```
This result doesn't look right

macwoody replied:
5728432093.013750076293945313

xhfanhe commented on test/JDBC/expected/BABEL-736.out:
```diff
@@ -0,0 +1,69 @@
+GO
+~~ERROR (Code: 8115)~~
+
+~~ERROR (Message: integer out of range)~~
```
This is different from SQL server output, is this expected?
```
1> SELECT POWER(CAST (241289984 AS BIGINT),2)
2> go

--------------------
   58220856378720256

(1 rows affected)
```

xhfanhe commented on test/JDBC/expected/BABEL-736.out:
```diff
@@ -0,0 +1,69 @@
+GO
+~~START~~
+numeric
+58220856378720256.0000000000000000
```
This is different from SQL server output, is this expected?
```
1> SELECT POWER(CAST (241289984 AS BIGINT),2.0)
2> go

--------------------
   58220856378720256

(1 rows affected)
```

xhfanhe commented on test/JDBC/input/BABEL-736.sql:
```diff
@@ -0,0 +1,26 @@
+GO
+
+SELECT POWER(CAST (241289984 AS BIGINT),2.0)
+GO
```
We need upgrade tests for newly added functions.

pratikzode committed:
DEGREES, RADIANS and POWER return float always

Changed int to BIGINT in DEGREES, RADIANS and POWER.

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
DEGREES, RADIANS and POWER return float always
Have added the test cases

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1191,61 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+ {
+    int64    arg1 = PG_GETARG_INT64(0);
+    int64    arg2 = PG_GETARG_INT64(1);
+    float8  result;
```
align the indentation

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4120,6 +4120,19 @@ ALTER FUNCTION sys.get_max_id_from_table RENAME TO get_max_id_from_table_depreca
+CREATE OR REPLACE FUNCTION sys.degrees(IN arg1 BIGINT)
+RETURNS int  AS 'babelfishpg_tsql','int_degrees' LANGUAGE C STRICT IMMUTABLE PARALLEL SAFE;
+GRANT EXECUTE ON FUNCTION sys.degrees(BIGINT) TO PUBLIC;
+=======
```
remove these merge conflicts markers

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4120,6 +4120,19 @@ ALTER FUNCTION sys.get_max_id_from_table RENAME TO get_max_id_from_table_depreca
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_min_id_from_table_deprecated_in_2_3_0');
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_max_id_from_table_deprecated_in_2_3_0');
 
+<<<<<<< HEAD
```
delete merge conflict markers

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4175,6 +4188,7 @@ $$ LANGUAGE plpgsql IMMUTABLE STRICT;
 
 -- Drop the deprecated function
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_tds_id_deprecated_2_3_0');
+>>>>>>> upstream/BABEL_2_X_DEV
```
same as above

macwoody commented on test/JDBC/expected/BABEL-736-vu-prepare.out:
```diff
@@ -0,0 +1,39 @@
+CREATE TABLE BABEL_736_vu_prepare_t_1(
+		a int,
+		b float,
+		c int NOT NULL,
```
please remove not null and unique and focus on normal values

pratikzode committed:
DEGREES, RADIANS and POWER return float always

Have added INT,BIGINT,smallint,tinyint to the degrees function

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1194,127 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+}
+
+Datum
+BIGINT_degrees(PG_FUNCTION_ARGS)
```
plz use lower case for bigint

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1194,127 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+    	result = floor(result);
+
+	 /* Range check */
+	if (unlikely(isnan(result) || !FLOAT8_FITS_IN_INT16(result)))
```
Here, you may need to check whether it fits int 1Byte, which is definition of tinyint

macwoody commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -3231,6 +3231,30 @@ RETURNS sys.NVARCHAR(128)  AS 'babelfishpg_tsql' LANGUAGE C;
+GRANT EXECUTE ON FUNCTION sys.degrees(BIGINT) TO PUBLIC;
+
+CREATE OR REPLACE FUNCTION sys.degrees(IN arg1 INT)
+RETURNS int AS 'babelfishpg_tsql','int_degrees' LANGUAGE C STRICT IMMUTABLE PARALLEL SAFE;
```
one of the above should be bigint

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4120,6 +4120,17 @@ ALTER FUNCTION sys.get_max_id_from_table RENAME TO get_max_id_from_table_depreca
+CREATE OR REPLACE FUNCTION sys.degrees(IN arg1 BIGINT)
+RETURNS int  AS 'babelfishpg_tsql','int_degrees' LANGUAGE C STRICT IMMUTABLE PARALLEL SAFE;
+GRANT EXECUTE ON FUNCTION sys.degrees(BIGINT) TO PUBLIC;
 
```
same definition should be added here as well.

pratikzode committed:
Defined a FLOAT8_FITS_IN_INT8 for tinyint

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
DEGREES, RADIANS and POWER return float always

In Babel, the Transact-SQL functions DEGREES, RADIANS and POWER always return float instead of returning the same type.
Now Transact-SQL functions DEGREES, RADIANS and POWER is returning the
same type as in case sql server.

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 DEGREES, RADIANS and POWER return float always

Defined mininum and maximum for tinyint

Task:BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Update on test cases

Updated the test cases and has defined int8 for tinyint

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 Updated the merge conflict

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1194,134 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+    PG_RETURN_INT16((int16)result);
+}
+
+#define PG_INT8_MIN		(-0x7F-1)
```
These two seem already exist in PG , don't have to redefine them.

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1194,134 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+
+#define PG_INT8_MIN		(-0x7F-1)
+#define PG_INT8_MAX		(0x7F)
+#define PG_GETARG_INT8
```
what does this macro do ?

pratikzode replied:
#define PG_GETARG_INT8(n)	 DatumGetInt8(PG_GETARG_DATUM(n))
Macros for fetching arguments of standard types

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1194,134 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("integer out of range")));
+
+    PG_RETURN_INT16((int16)result);
```
could you please adjust the indentation ?

macwoody commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -3231,6 +3231,34 @@ RETURNS sys.NVARCHAR(128)  AS 'babelfishpg_tsql' LANGUAGE C;
+RETURNS int AS 'babelfishpg_tsql','int_degrees' LANGUAGE C STRICT IMMUTABLE PARALLEL SAFE;
+GRANT EXECUTE ON FUNCTION sys.degrees(INT) TO PUBLIC;
+
+CREATE OR REPLACE FUNCTION sys.degrees(IN arg1 smallint)
```
for the datatype, some are in lower case, like this line, some are in upper case. 
please choose one style

pratikzode committed:
BABEL-736 alignment update on .c file

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 Added bigint, int, smallint, tinyint ,int to degrees

Have a bigint, int, smallint, tinyint ,int to degrees which is giving same output as SQL server.
Have added the test cases for each function separately for better understanding.

Task: BABEL-736

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 Alignment in Function.c file

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 Added MVU test cases

Have a bigint, int, smallint, tinyint ,int in the test cases

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -76,6 +77,14 @@ PG_FUNCTION_INFO_V1(language);
 PG_FUNCTION_INFO_V1(procid);
 PG_FUNCTION_INFO_V1(babelfish_integrity_checker);
+PG_FUNCTION_INFO_V1(bigint_power);
+PG_FUNCTION_INFO_V1(int_power);
```
please remove all power, radian implementations from this PR

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1196,167 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+	if (unlikely(isnan(result) || !FLOAT8_FITS_IN_INT32(result)))
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				 errmsg("Arithmetic overflow error for data type smallint")));
```
can this be triggered in theory?

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1196,167 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Arithmetic overflow error for data type tinyint")));
+
+	PG_RETURN_INT16((int16)result);
```
return int32

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1196,167 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+	if (unlikely(isnan(result) || !FLOAT8_FITS_IN_INT16(result)))
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Arithmetic overflow error for data type tinyint")));
```
can this be triggered in theory?

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1196,167 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+	PG_RETURN_INT32((int32) result);
+}
+
+#define DatumGetInt8(X) ((int8) (X))
```
please remove unused codes

macwoody commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -3231,6 +3231,38 @@ RETURNS sys.NVARCHAR(128)  AS 'babelfishpg_tsql' LANGUAGE C;
 CREATE OR REPLACE FUNCTION sys.host_name()
 RETURNS sys.NVARCHAR(128)  AS 'babelfishpg_tsql' LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
+CREATE OR REPLACE FUNCTION sys.power(IN arg1 BIGINT, IN arg2 BIGINT)
```
keep only degrees functions

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4120,6 +4120,37 @@ ALTER FUNCTION sys.get_max_id_from_table RENAME TO get_max_id_from_table_depreca
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_min_id_from_table_deprecated_in_2_3_0');
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_max_id_from_table_deprecated_in_2_3_0');
 
+CREATE OR REPLACE FUNCTION sys.power(IN arg1 BIGINT, IN arg2 BIGINT)
```
keep only degrees function

macwoody commented on test/JDBC/expected/BABEL-736-vu-prepare.out:
```diff
@@ -0,0 +1,50 @@
+GO
+~~ROW COUNT: 1~~
+
+INSERT INTO BABEL_736_vu_prepare_t_1 VALUES (2147483648, 20.1, 64213, NULL, 20);
```
Let's discuss

macwoody commented on test/JDBC/expected/BABEL-736-vu-verify.out:
```diff
@@ -0,0 +1,21 @@
+~~END~~
+
+
+SELECT * from BABEL_736_vu_prepare_t_1 ORDER BY 3
```
why ORDER 3 ?

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1196,167 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+Datum
+smallint_degrees(PG_FUNCTION_ARGS)
+{
+	int32	arg1 = PG_GETARG_INT32(0);
```
INT16

macwoody replied:
this one not fixed

pratikzode committed:
BABEL-736 Just added Degrees(bigint, int, smallint, tinyint)

Have removed radians and power and have added only degrees.
In degrees have added new bigint, int, smallint, tinyint function and
Tested it as well.

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on test/JDBC/expected/BABEL-736-vu-verify.out:
```diff
@@ -0,0 +1,14 @@
+SELECT * from BABEL_736_vu_prepare_t_1 ORDER BY 1
```
SELECT degrees(a), degrees(b), degrees(c), degress(d)... from ..

macwoody commented on test/JDBC/expected/BABEL-736.out:
```diff
@@ -0,0 +1,245 @@
+~~ERROR (Message: Arithmetic overflow error converting expression to data type int)~~
+
+
+SELECT degrees(9999.0 * 9999.0)
```
this result is different from sql server, and is not part of your implementation, please remove this case

macwoody commented on test/JDBC/expected/BABEL-736.out:
```diff
@@ -0,0 +1,245 @@
+5.72843209301375E9
+~~END~~
+
+
```
please also add one case , where the numeric value fits in bigint range

macwoody commented on test/JDBC/expected/BABEL-736.out:
```diff
@@ -0,0 +1,245 @@
+GO
+~~ERROR (Code: 220)~~
+
+~~ERROR (Message: smallint out of range)~~
```
why do you still see this error ? 
Please make sure your test could run without problem

macwoody commented on test/JDBC/expected/BABEL-736.out:
```diff
@@ -0,0 +1,245 @@
+bigint
+~~ERROR (Code: 8115)~~
+
+~~ERROR (Message: bigint out of range)~~
```
same here

pratikzode committed:
BABEL-736 Updated the test for degrees

Have updated the test for degrees, changed to int16 in smallint_degrees

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 Updated the test cases

Have updated the test for degrees in BABEL-736.out,
BABEL-736-vu-verify.out

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on test/JDBC/expected/BABEL-736-vu-verify.out:
```diff
@@ -0,0 +1,14 @@
+SELECT degrees(a), degrees(b), degrees(c), degrees(d), degrees(e) from BABEL_736_vu_prepare_t_1 
```
order by is mandatory for all the select statement returning multiple rows in test

AvantikaDG replied:
+1

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1192,83 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+	else
+		result = floor(result);
+
+	/* Range check it cannot overflow int32 */
```
/* skip range check, since it cannot overflow int32 */

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1192,83 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+		result = ceil(result);
+	else
+		result = floor(result);
+
```
same here like above

pratikzode committed:
BABEL-736 Added a range check comment

Have added comment /* skip range check, since it cannot overflow int32 */ smallint, tinyint.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 Updated the upgrade script

In the upgrade script it was returning int
 so have to change it to bigint in babelfish_tsql—2.2.0—2.3.0.sql

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-736 Update expected_dependency.out 

Have added tests for these 4 things  
+Function sys.degrees(bigint)
+Function sys.degrees(integer)
+Function sys.degrees(smallint)
+Function sys.degrees(sys.tinyint)

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

AvantikaDG commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1192,85 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+	else
+		result = floor(result);
+
+	 /* Range check */
```
Since both int64 and float8 are 8 bytes, please add boundary test cases for max positive and negative values of bigint.

AvantikaDG commented on test/JDBC/input/BABEL-736.sql:
```diff
@@ -0,0 +1,95 @@
+SELECT degrees(10)
```
Where is this test running? I don't see it in the jdbc_schedule. Is this intentional?

pratikzode replied:
yes it was given in the jira so have added it

AvantikaDG replied:
I mean this test file.. where is this test file running?

AvantikaDG commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1192,85 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+		result = floor(result);
+
+	 /* Range check */
+	if (unlikely(isnan(result) || !FLOAT8_FITS_IN_INT32(result)))
```
Do we have a test case that can exercise this code?

pratikzode replied:
SELECT degrees(CAST(2147483648 AS INT));

2jungkook replied:
The above example throws an error in `CAST(2147483648 AS INT)` part, not in this function.

pratikzode committed:
BABEL-736 Updated the Test Case

Have added the order by clause in verify file.

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

2jungkook commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql:
```diff
@@ -4120,6 +4120,21 @@ ALTER FUNCTION sys.get_max_id_from_table RENAME TO get_max_id_from_table_depreca
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_min_id_from_table_deprecated_in_2_3_0');
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'get_max_id_from_table_deprecated_in_2_3_0');
 
+CREATE OR REPLACE FUNCTION sys.degrees(IN arg1 BIGINT)
```
Please use `2.3.0--2.4.0` upgrade script.

2jungkook commented on test/JDBC/upgrade/13_4/schedule:
```diff
@@ -101,6 +101,7 @@ Test-sp_helpsrvrolemember
 BABEL-404
 BABEL-493
 BABEL-621
+BABEL-736
```
Do we really want to test previous versions' behavior? If so, follow this (https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/270/files), otherwise please put this only in the latest schedule file.

2jungkook commented on test/python/expected/upgrade_validation/expected_dependency.out:
```diff
@@ -359,6 +359,10 @@ Function sys.db_name()
+Function sys.degrees(bigint)
+Function sys.degrees(integer)
+Function sys.degrees(smallint)
+Function sys.degrees(sys.tinyint)
```
We should not add entries in this file.

2jungkook commented on test/JDBC/input/BABEL-736.sql:
```diff
@@ -0,0 +1,95 @@
+SELECT degrees(10)
```
1. Please move out everything in this file into the upgrade test and use proc/view there (ref: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/876)
2. We should have all tests where input parameters are any data types that can be implicitly casted into numeric expression, including const number and literal.

2jungkook commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1192,85 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+}
+
+Datum
+tinyint_degrees(PG_FUNCTION_ARGS)
```
`smallint_degrees` and `tinyint_degrees` are duplicate?

2jungkook commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1187,3 +1192,85 @@ babelfish_integrity_checker(PG_FUNCTION_ARGS)
+	if (unlikely(isnan(result) || !FLOAT8_FITS_IN_INT64(result)))
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Arithmetic overflow error converting expression to data type bigint")));
```
There is no test hitting this line.

2jungkook commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -18,6 +18,7 @@
 #include "tsearch/ts_locale.h"
 #include "utils/acl.h"
 #include "utils/builtins.h"
+#include "utils/datetime.h"
```
why do we need this?

2jungkook commented:
Please update the PR title and description accordingly.

pratikzode committed:
BABEL-736 All alignment update

Task: BABEL-736
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-3747 Bigint/Int/Smallint/Tinyint Degrees Function

This Jira is a part of BABEL-736.
At first in Babel, the Transact-SQL functions degrees always return float instead of returning the same type in SQL server.
In this Jira we have added Bigint, Int, Smallint and Tinyint to degrees function.
Have added the test case to show Bigint, Int, Smallint and Tinyint to degrees.
Now this 3 is giving same output as SQL server.

Task: BABEL-3747
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-3747 Updated the babelfishpg_tsql--2.3.0--2.4.0.sql

Task: BABEL-3747
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-3747 Deleted the previous test cases from JDBC/expected

Task: BABEL-3747
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-3747 Added View test cases for Bigint, Int, Smallint and Tinyint

Task: BABEL-3747
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Update babelfishpg_tsql--2.2.0--2.3.0.sql
